src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -520,6 +521,42 @@ private BoundExpression CreateAndPopulateArray(BoundCollectionExpression node, A
                 throw ExceptionUtilities.UnexpectedValue(node);
             }
 
+            // Collection-expr is of the form `[..spreadExpression]`, where 'spreadExpression' has same element type as the target collection.
+            // Optimize to `spreadExpression.ToArray()` if possible.
+            if (node is { Elements: [BoundCollectionExpressionSpreadElement { Expression: { } spreadExpression } spreadElement] }
+                && spreadElement.IteratorBody is BoundExpressionStatement expressionStatement
+                && expressionStatement.Expression is not BoundConversion)
+            {
+                var spreadTypeOriginalDefinition = spreadExpression.Type!.OriginalDefinition;
+                if (tryGetToArrayMethod(spreadTypeOriginalDefinition, WellKnownType.System_Collections_Generic_List_T, WellKnownMember.System_Collections_Generic_List_T__ToArray) is { } listToArrayMethod)
+                {
+                    var rewrittenSpreadExpression = VisitExpression(spreadExpression);
+                    return _factory.Call(rewrittenSpreadExpression, listToArrayMethod.AsMember((NamedTypeSymbol)spreadExpression.Type!));
+                }
+
+                if (TryGetSpanConversion(spreadExpression.Type, out var asSpanMethod))
+                {
+                    var spanType = CallAsSpanMethod(spreadExpression, asSpanMethod).Type!.OriginalDefinition;
+                    if ((tryGetToArrayMethod(spanType, WellKnownType.System_Span_T, WellKnownMember.System_Span_T__ToArray)
+                            ?? tryGetToArrayMethod(spanType, WellKnownType.System_ReadOnlySpan_T, WellKnownMember.System_ReadOnlySpan_T__ToArray))
+                        is { } toArrayMethod)
+                    {
+                        var rewrittenSpreadExpression = CallAsSpanMethod(VisitExpression(spreadExpression), asSpanMethod);
+                        return _factory.Call(rewrittenSpreadExpression, toArrayMethod.AsMember((NamedTypeSymbol)rewrittenSpreadExpression.Type!));
+                    }
+                }
+
+                MethodSymbol? tryGetToArrayMethod(TypeSymbol spreadTypeOriginalDefinition, WellKnownType wellKnownType, WellKnownMember wellKnownMember)
```

Comment:
This isn't really a try-get in traditional form, and it feels like the code that uses it is a bit convoluted because of it. Consider refactoring to be a standard `tryGet` with an `out` parameter.